ops/framework.py

```diff
@@ -633,20 +709,20 @@ def commit(self):
         self.save_snapshot(self._stored)
         self._storage.commit()
 
-    def register_type(self, cls, parent: typing.Optional["_ParentHandle"], kind=None):
+    def register_type(self, cls: 'Type[_Serializable]', parent: Optional['_ParentHandle'],
+                      kind: str = None):
         """Register a type to a handle."""
         if parent is not None and not isinstance(parent, Handle):
+            assert isinstance(parent, Object), parent  # type guard
             parent = parent.handle
-        if parent:
-            parent_path = parent.path
-        else:
-            parent_path = None
-        if not kind:
-            kind = cls.handle_kind
+
+        assert isinstance(parent, (type(None), Handle)), parent  # type guard
+        parent_path = parent.path if parent else None  # type: Optional[str]
```

Comment:
I really think something like this would be more clear here:

```python
        parent_path = None  # type: Optional[str]
        if isinstance(parent, Object):
            parent_path = parent.handle.path
        elif isinstance(parent, Handle):
            parent_path = parent.path
```